src/fairseq2/datasets/data_reader.py

```diff
@@ -107,12 +107,17 @@ def __next__(self) -> Tuple[int, List[BatchT]]:
 
             batches.append(batch)
 
+        self._eod = len(batches) != self._num_accumulate
+
         # If requested, sync batches across all processes in the gang.
-        if self._sync_batches:
-            total_batch_size, self._eod = _all_batch_sizes(batches, self._gang, logger)
+        if self._sync_batches and self._gang.size > 1:
+            total_batch_size, self._eod = _total_batch_size_and_eod(
+                batches, self._eod, self._gang, logger
+            )
         else:
-            total_batch_size = self._gang.size * self._num_accumulate * batch.batch_size
-            self._eod = len(batches) != self._num_accumulate
+            total_batch_size = self._gang.size * sum(
```

Comment:
This statement, while correct, is a little confusing but it implies we might be double counting the batch size due to use of  both`self._gang_size`  and `sum(...batches)`. This is correct since `self._gang_size` is 1 for this codepath (I assume 0 is not a valid value). If that is the case, we could consider simplifying this statement by dropping `self._gang.size`

Reply:
`self._gang_size` is not forced to be 1 for this codepath, it can be > 1 and `self._sync_batches` can be False. In that can all ranks have the same number of batches which is `sum(batch.batch_size for batch in batches)`, so we do have to multiply by `self._gang_size`. To clarify, when we iterate over `batches` we iterate over `range(self._num_accumulate)` (due to gradient accumulation), so this for loop doesn't take into account the different ranks.

Reply:
Thank you for the context!